Common/WeakObserverSet.swift

```diff
@@ -0,0 +1,48 @@
+//
+//  WeakObserverSet.swift
+//  RileyLink
+//
+//  Created by Pete Schwamb on 11/18/18.
+//  Copyright © 2018 Pete Schwamb. All rights reserved.
+//
+
+import Foundation
+
+public struct WeakObserverSetIterator<Observer>: IteratorProtocol {
+    let enumerator: NSEnumerator
+    
+    init(_ enumerator: NSEnumerator) {
+        self.enumerator = enumerator
+    }
+    
+    public mutating func next() -> Observer? {
+        return enumerator.nextObject() as? Observer
+    }
+}
+
+public class WeakObserverSet<Observer>: Sequence {
```

Comment:
This is an interesting approach to tackling the problem of aggregating weak references; I wasn't aware of `NSHashTable.weakObjects()`.
However, we can implement this generically without requiring NSObject inheritance (or relying on Obj-C classes at all). Note that it's impossible to retain complete type safety when working with class-constrained protocol types as the generic parameter, as protocols don't conform to themselves*, which is the issue we're fighting here. I [wrote about experiencing this problem](https://forums.swift.org/t/using-class-constrained-protocol-as-anyobject/13988) a few months ago but didn't get much solid feedback.

That said, [here's a sketch](https://gist.github.com/mpangburn/d8b0d51b688fe6ffb10ef8d44dd15b85) of a pure-Swift WeakSet type. It still suffers the limitation of not being able to enforce that its generic parameter is a class or class-constrained protocol at compile-time, but I think it's an improvement over the Obj-C-reliant variant.

* There are a couple exceptions here: `@objc`-annotated protocols self-conform, and there's [an open PR in apple/swift](https://github.com/apple/swift/pull/20629) to do the same for Error in light of recent discussion regarding the addition of Result to the stdlib. Adding a way to self-conform a protocol in Swift has come up but doesn't have concrete plans AFAIK.